src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs

```diff
@@ -90,68 +78,54 @@ public void InvokeCompletionCallback(uint sniErrorCode)
         /// Allocate space for data
         /// </summary>
         /// <param name="capacity">Length of byte array to be allocated</param>
-        public void Allocate(int capacity)
+        public void Allocate(int capacity, bool reserveMuxHeader)
         {
-            if (_data != null && _data.Length < capacity)
+            SNIPacketFlags flags = reserveMuxHeader ? SNIPacketFlags.MuxHeaderReserved : SNIPacketFlags.None;
+            int headerCapacity = reserveMuxHeader ? SNISMUXHeader.HEADER_LENGTH : 0;
+            int totalCapacity = headerCapacity + capacity;
+            if (_data != null)
             {
-                if (_isBufferFromArrayPool)
+                if (_data.Length < totalCapacity)
                 {
-                    ArrayPool<byte>.Shared.Return(_data);
+                    Array.Clear(_data, 0, _header + _length);
+                    if ((_flags & SNIPacketFlags.ArrayFromPool) == SNIPacketFlags.ArrayFromPool)
+                    {
+                        ArrayPool<byte>.Shared.Return(_data, clearArray: false);
+                        _flags &= ~SNIPacketFlags.ArrayFromPool;
+                    }
+                    _data = null;
+                }
+                else
+                {
+                    // if the current array is big enough and rented keep it
+                    flags |= (_flags & SNIPacketFlags.ArrayFromPool);
                 }
-                _data = null;
             }
 
             if (_data == null)
             {
-                _data = ArrayPool<byte>.Shared.Rent(capacity);
-                _isBufferFromArrayPool = true;
+                _data = ArrayPool<byte>.Shared.Rent(totalCapacity);
+                flags |= SNIPacketFlags.ArrayFromPool; // set local not instance because it will be assigned after this block
             }
 
+            _flags = flags;
             _capacity = capacity;
```

Comment:
It makes more sense to have _capacity = totalCapacity instead of capacity. 
In case the totalCapacity has the header length in it, it is certain that MARS Handle will write the header before writing out the packet that it needs to send. And this way you wont have to change the capacity in the `SetHeader` method below.